src/lib.rs

```diff
@@ -93,13 +94,74 @@ pub use crate::char_data::{bidi_class, HardcodedBidiData};
 use alloc::borrow::Cow;
 use alloc::string::String;
 use alloc::vec::Vec;
+use core::char;
 use core::cmp;
 use core::iter::repeat;
 use core::ops::Range;
+use core::str::CharIndices;
 
 use crate::format_chars as chars;
 use crate::BidiClass::*;
 
+/// Trait that abstracts over a text source for use by the bidi algorithms.
+/// We implement this for str (UTF-8) and for [u16] (UTF-16, native-endian).
+/// (For internal unicode-bidi use; API may be unstable.)
+/// This trait is sealed and cannot be implemented for types outside this crate.
+pub trait TextSource<'text>: private::Sealed
+where
+    <Self as TextSource<'text>>::CharIndexIter: Iterator<Item = (usize, char)>,
```

Comment:
not: not in the where bounds, they can go on the associated types directly: `type CharIndexIterator: Iterator<...>`